### ISSUE TITLE
binary_semaphore::max() requirements

### ISSUE THREAD

**Issue by hadrielk** [brycelelbach/wg21_p1135_cpp20_synchronization_library#12]
Hi,
Since `binary_semaphore` is an alias for `counting_semaphore<1>`, there are some weird implications that I wanted to confirm were intentional (or at least not undesired).

`counting_semaphore::max()` says this:
```
Returns: The maximum value of counter. This value is greater than or equal to least_max_value.
```
which means a `binary_semaphore::max()` can theoretically return a value greater than `1`.

In the original proposal [P0514R4](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0514r4.pdf), however, it implied the max value was required to be `1`.  (I think?)

The only reason this matters, and why I'm asking, is whether a user can _always_ safely do this to initialize the `binary_semaphore` via the constructor:
```
std::binary_semaphore sem{std::binary_semaphore::max()};
```

Per the current wording of [N4842](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/n4842.pdf), doing that would be valid, even if `binary_semaphore:max()` yields `2`. But presumably that would break the semantics of `binary_semaphore`.

And the same issue applies for `binary_semaphore::release(binary_semaphore::max())`, although in practice presumably users would never provide an argument for `binary_semaphore::release()` since the default argument value is `1` anyway.

----

The reason this stuff matters, I think, is that users could write generic template classes/functions that only deal with `counting_semaphore<N>` by using `max()` generically, without having to specialize for `binary_semaphore`/`counting_semaphore<1>`... if it were safe to do so.

----

As an aside, arguably this `max()` issue applies to `counting_semaphore`s in general - it seems odd to me that `max()` could return a value greater than the template param value. The user _specified_ the max value using the template param, and that's what they should always see; the fact that internally it might be larger should be a hidden detail, no? Otherwise using `max()` would lead to exceeding the specified template value.

Having said that, I recognize the non-type template param is specifically named `least_max_value`, and thus provides the hint it's only a _minimum_ max value. So if I want to say: "up to 4 concurrent accessors are allowed", I can't. I can only say: "at least 4 or more concurrent accessors are allowed". Is that the intention?

**Comment by brycelelbach** [comment]
Adding my coauthors

**Comment by ogiroux** [comment]
This is a correct read of the consequences of the wording, and I think that this is largely intentional via an intentional choice to make binary and counting semaphores the same template type in the final library design.

Initializing a semaphore to the implementation-defined maximum value, without checking it and instead of a value derived from the application's domain, is a pretty solid anti-pattern. It's true that you could do that, and you could do that in generic code, but I think the value of keeping binary and counting semaphores together outweighs the inconvenience of writing the anti-pattern in generic code. Someone might even say that making anti-patterns inconvenient to write is good (typically that's not my view, but I'm channeling others here).

Note the "up to 4 concurrent accessors" usage is 100% the core purpose of the semaphore itself, modeled via the dynamic initialization value argument. The static template argument's only purpose is to ensure that, if you were to initialize it with 4 at run-time, then you can let the implementation know at compile-time that you're going to do that so it won't choose its internal implementation that tops out at 3 but is super fast. So what you want there is soundly supported, it's the whole point of this facility even.

In short, the semaphores main reason for existing is well served by the design; some fringe use-cases are less so. I doubt we would make a design change here, but we can always entertain a paper to restrict the implementation-defined maximum of binary semaphore to always be 1. That would be an ABI break for implementations so it's not likely to pass, but we would discuss it.

**Comment by hadrielk** [comment]
Oh, I'm not asking for a change at all - just clarification of intent. I'm creating the pages for it for [cppreference.com](https://en.cppreference.com), and just wanted to understand the design choices/intent.

In my mind I was thinking of `max()` as a convenience constant for users of the class, to be usable without having to know what an actual type's `least_max_value` was. (for example so one could do `using semaphore = std::counting_semaphore<4>;` in some common/library header, and users of it could do `semapahore::max()` for initialization without having to know `4`) But really this is a very minor thing, and I have no idea if anyone would ever want to do such a thing, and it's easily avoidable using an `integral_constant` or some such anyway. I was more just slightly surprised/confused, really.

Thanks for the reply!
